File: src/worker.ts
import type {Database, InitSqlJsStatic} from "sql.js";
import initSqlJs from "@jlongster/sql.js";
import {SQLiteFS} from "absurd-sql";
import IndexedDBBackend from "absurd-sql/dist/indexeddb-backend";
import {expose} from "comlink";

let _db: Database | null = null;
function db() {
  if (_db == null) {
    throw new Error("not initialized");
  }
  return _db;
}

async function setupSqljs(dbPath: string) {
  const SQL = await (initSqlJs as InitSqlJsStatic)({
    locateFile: (file: string) => file,
  });
  // @ts-ignore
  const sqlFS = new SQLiteFS(SQL.FS, new IndexedDBBackend());
  // @ts-ignore
  SQL.register_for_idb(sqlFS);
  // @ts-ignore
  SQL.FS.mkdir("/sql");
  // @ts-ignore
  SQL.FS.mount(sqlFS, {}, "/sql");
  class PatchedDatabase extends SQL.Database {
    constructor() {
      // @ts-ignore
      super(dbPath, { filename: true });
      // Set indexeddb page size
      this.exec(`PRAGMA page_size=8192;PRAGMA journal_mode=MEMORY;`);
    }
  }
  _db = new PatchedDatabase();
}
async function setup() {
  // with /sql/ namespace
  const DBNAME = "/sql/db.sqlite";
  await setupSqljs(DBNAME);
  // conn = await createConnection({
  //   type: "sqljs", // this connection search window.SQL on browser
  //   location: DBNAME,
  //   autoSave: false, // commit by absurd-sql
  //   synchronize: true,
  //   entities: [User],
  //   logging: ["query", "schema"],
  // });
  // userRepository = conn.getRepository(User);
}

async function execRaw(query: string) {
  return db().exec(query);
}
async function run(query: string, args: any[] = []) {
  if (args.length === 0) {
    const stmt = db().prepare(`SELECT SUM(value) FROM kv`);
    stmt.step();
    const result = stmt.getAsObject();
    stmt.free();
    return result;
  } else {
    const stmt = db().prepare(query);
    stmt.run(args);
    const result = stmt.getAsObject();
    stmt.free();
    return result;
  }
}

async function runMany(query: string, argsList: Array<Array<any>> = []) {
  db().exec("BEGIN TRANSACTION");
  const stmt = db().prepare(query);
  const results = argsList.map((args) => {
    stmt.run(args);
    return stmt.getAsObject();
  });
  stmt.free();
  db().exec("COMMIT");
  return results;
}

const api = { setup, run ,execRaw, runMany };
export type Api = typeof api;

expose(api);

File: src/Worker.fs
module Worker

open System
open Fable.Core
open Fable.Core.JsInterop

//let [<Import("initSqlJs", from="@jlongster/sql.js")>] initSqlJsStatic: SqlJs.InitSqlJsStatic = jsNative

//let getTheme(x: int): IInterface = importDefault "my-lib"
// JS: import getTheme from "my-lib"

let initSqlJs: SqlJs.InitSqlJsStatic = importDefault "@jlongster/sql.js"
//type [<AllowNullLiteral>] SQLiteFS = 
//    [<Emit "new $0($1...)">] abstract Constructor: obj * obj -> obj
    
let [<Import("SQLiteFS", from="absurd-sql")>] sqlitefs: obj = jsNative

//let [<Import("IndexedDBBackend", from="absurd-sql/dist/indexeddb-backend")>] IndexedDBBackend: obj = jsNative
//type IndexedDbBackendType =
//   [<Emit "new $0($1...)">] abstract Constructor: unit -> obj

let IndexedDBBackend: JsConstructor = importDefault "absurd-sql/dist/indexeddb-backend"


let mutable _db : SqlJs.Database option = None

let db() =
   match _db with
   | None -> failwith "not initialized"
   | Some db -> db


type SqlJsConfig() =
   interface SqlJs.SqlJsConfig with
      override _.locateFile(s,_)= s
   
let setupSqlJs (dbPath:string) = promise {
   let! SQL = initSqlJs.Invoke(SqlJsConfig())
   
   let sqlFs = createNew sqlitefs (SQL.FS, IndexedDBBackend.Create())
   
   SQL.register_for_idb(sqlFs)
   
   SQL.FS?mkdir("/sql")
   SQL.FS?mount(sqlFs, {||},"/sql")
   let db: SqlJs.Database = createNew SQL.Database (dbPath, {|filename=true|}) :?> SqlJs.Database
   db?exec("PRAGMA page_size=8192;PRAGMA journal_mode=MEMORY;PRAGMA mmap_size=0;PRAGMA synchronous=NORMAL;PRAGMA temp_store=MEMORY")
   _db <- Some db
}

let setup() = promise {
    // with /sql/ namespace
    let dbname = "/sql/db.sqlite"
    do! setupSqlJs(dbname)
}


let execRaw (query: string) = promise {
   return db().exec(query)
}

let runMany (query:string) argsList =
   promise {
      db().exec("BEGIN TRANSACTION") |> ignore
      let stmt = db().prepare(query)
      let results = argsList
                    |> Array.map (fun args ->
                       stmt.run(!!args)
                       stmt.getAsObject()
                    )
      stmt.free() |> ignore
      db().exec("COMMIT") |> ignore
      return results
   }

let runPreparedStep(query: string) = promise {
   let stmt = db().prepare(query)
   stmt.step() |> ignore
   let result = stmt.getAsObject()
   stmt.free() |> ignore
   return result
}


type IApi =
   abstract member execRaw: string -> JS.Promise<ResizeArray<SqlJs.QueryExecResult>>
   abstract member setup: unit -> JS.Promise<unit>
   
   abstract member runMany: string -> obj [] -> JS.Promise<SqlJs.ParamsObject []>
   abstract member runPreparedStep: string-> JS.Promise<SqlJs.ParamsObject>
   
let api = createObj [
                      "setup" ==> setup
                      "execRaw" ==> execRaw
                      "runMany" ==> runMany
                      "runPreparedStep" ==> runPreparedStep
                      ]

Comlink.expose(api)
   
   